lib/chkhash.c

```diff
@@ -0,0 +1,70 @@
+#include <config.h>
+
+#include "chkhash.h"
+
+#include <regex.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <string.h>
+
+/*
+ * match_regex - return true if match, false if not
+ */
+bool 
+match_regex(const char *pattern, const char *string) 
+{
+	regex_t regex;
+	int result;
+
+	if (regcomp(&regex, pattern, REG_EXTENDED) != 0)
+		return false;
+
+	result = regexec(&regex, string, 0, NULL, 0);
+	regfree(&regex);
+
+	return result == 0;
+}
+
+/*
+ * is_valid_hash - check if the given string is a valid password hash
+ *
+ * Returns true if the string appears to be a valid hash, false otherwise.
+ */
+bool 
+is_valid_hash(const char *hash) 
+{
+	if (hash == NULL)
+		return false;
+
+	// Minimum length for DES
+	if (strlen(hash) < 13)
+		return false;
+
+	// DES: exactly 13 characters from [A-Za-z0-9./]
+	if (match_regex("^[A-Za-z0-9./]{13}$", hash))
+		return true;
+
+	// MD5: $1$ + salt + $ + 22-char hash
+	if (match_regex("^\\$1\\$[A-Za-z0-9./]{1,16}\\$[A-Za-z0-9./]{22}$", hash))
+		return true;
+
+	// SHA-256: $5$ + salt + $ + 43-char hash
+	if (match_regex("^\\$5\\$[A-Za-z0-9./]{1,16}\\$[A-Za-z0-9./]{43}$", hash))
+        	return true;
+
+	// SHA-512: $6$ + salt + $ + 86-char hash
+	if (match_regex("^\\$6\\$[A-Za-z0-9./]{1,16}\\$[A-Za-z0-9./]{86}$", hash))
+		return true;
+
+	// Bcrypt: $2[abzy]$ + 2-digit cost + $ + 53-char hash
+	if (match_regex("^\\$2[abxy]\\$[0-9]{2}\\$[A-Za-z0-9./]{53}$", hash))
+		return true;
+
+	// Yescrypt: $y$ + algorithm parameters + $salt + 53-char (minimum) hash
+	if (match_regex("^\\$y\\$[./A-Za-z0-9]+\\$[./A-Za-z0-9]{1,}\\$[./A-Za-z0-9]{43,}$", hash))
```

Comment:
For consistency with the rest, I'd say `$ + salt` instead of `$salt`.  And I think you forgot one `+ $ +` between the salt and the hash.